Contribution workflow PR (#9)
r-devel/r-dev-env: README.md

    <img src="screenshots/running_r_code.png" width="80%">
</p>

## R Contribution Workflow

### Build Setup
1. Environment Variables

>     - We have environment variables for setting path to install and build R and source code. And also for R source code version.

Reviewer: ```suggestion
    - We have environment variables for setting the paths for building R and storing the source code.
```

>     - The path ENV variable for R Build and R Source code are BUILDDIR and TOP_SRCDIR respectively. And for R source code version it is R_VERSION.

Reviewer 2: ```suggestion
    - The path ENV variable for R Build and R Source code are BUILDDIR and TOP_SRCDIR respectively..
```

    - The R_VERSION_BUILD_DIR has BUILDDIR path as base path and R_VERSION is added as suffix.
      
      ![image](https://github.com/r-devel/r-dev-env/assets/72031540/6b6cb661-cde3-4c69-845a-a6a281c18ec2)
2. svn checkout
   - The svn checkout cmd lets us create working of a repository with specific tag/branch.
   - Example:
    ```bash
     svn checkout https://svn.r-project.org/R/tags/$R_VERSION/ "$R_VERSION_SOURCE_DIR"
     ```
   - Here we are checking out R source with specific version and the checkout files and folders will be stored inside R_VERSION_SOURCE_DIR directory.
   - Output : We get file structure something like this after checking out R source code from R svn repository.
     ![image](https://github.com/r-devel/r-dev-env/assets/72031540/5c560a5e-a40d-4c68-9013-d6eec7327020)

3. cd to BUILDDIR
   - We need to change our directory to R build directory(BUILDDIR) to build and configure our R source code.
   - ```bash
     cd $R_VERSION_BUILD_DIR
     ```
   - The R_VERSION_BUILD_DIR has BUILDDIR path as base path and R_VERSION is added as suffix.
5. configure source code

>    - After we change directory to BUILDDIR we can build R and source code.

Reviewer: ```suggestion
   - After we change directory to BUILDDIR we can configure and build R.
```

   - ```bash
     "$R_VERSION_SOURCE_DIR/configure" --enable-R-shlib --without-recommended-packages
     make
     sudo make install
     ```

>    - The configure cmd help us build R and files and folders are stored inside BUILDDIR directory.

Reviewer: ```suggestion
   - The configure cmd prepares for building R, creating files and folders inside the BUILDDIR directory.
```

   - Output : We get file structure something like this after using configure command.
     
     ![image](https://github.com/r-devel/r-dev-env/assets/72031540/b6279710-1176-4c56-9a09-7c41c582e5f8)

> 6. After having configured setup, we can now make changes in source code and make our contributions.

Reviewer: ```suggestion
6. After having built the current development version of R, we can now make changes in source code and make our contributions.
```

### Contribution Workflow

> 1. R-org uses bugzilla for bug tracking and svn as a version control system.
> 2. We can find listed and known bugs on bugzilla. Link : https://bugs.r-project.org/buglist.cgi?bug_status=__open__
> 3. Example Contribution Workflow using DevContainer:
>    -  We will use bug [17616](https://bugs.r-project.org/show_bug.cgi?id=17616)

Reviewer: This is correct for my original suggestion (to show how to fix a bug in an old version of R). For the simpler workflow, we can delete this.

```suggestion
```

>    -  To confirm the bug we will click on `R:(not attach)` option which is in the bottom right of our R-dev codespace. It will open R terminal for us.

Reviewer: ```suggestion
   -  To start working in R we will click on `R:(not attach)` option which is in the bottom right of our R-dev codespace. It will open R terminal for us.
```

     
       ![image](https://github.com/r-devel/r-dev-env/assets/72031540/4ad3ed18-108a-4f29-ab6c-7f32d81721a7)
       ![image](https://github.com/r-devel/r-dev-env/assets/72031540/b3bdd3da-903d-4330-81c3-e41147d5dcd4)

>    -  We can now run R cmds. The expected output with bug (the second parameter label has the number of the factor level, "2", rather than the label of the factor level, "chilled") :
>       ```R
>       > lm(uptake~C(Treatment, contr.treatment), CO2)
>         Call:
>         lm(formula = uptake ~ C(Treatment, contr.treatment), data = CO2)
> 
>         Coefficients:
>                    (Intercept)  C(Treatment, contr.treatment)2  
>                          30.64                           -6.86
>       ```

Reviewer: New example:

```suggestion
   -  We can now run R commands. We will use the `utils::askYesNo()` function as an example
      ```R
      > askYesNo("Is this a good example?")
      Is this a good example? (Yes/no/cancel) Yes
      [1] TRUE
      ```
```

>  4. Edit the source code of stats::`contrasts<-()`  to implement the proposed fix. The source code can be found in $BUILDDIR/src/library/stats/R/contrasts.R. Before fix:

Reviewer: Edit the source code of `utils::askYesNo()` to change the default options
```suggestion
 4. Edit the source code of `utils::askYesNo()` to change the default options. The source code can be found in `$BUILDDIR/src/library/utils/R/askYesNo.R`. Before edit:
```

    ```R

>     if(is.function(value)) value <- value(nlevels(x))

Reviewer: ```suggestion
                     prompts = getOption("askYesNo", gettext(c("Yes", "No", "Cancel"))),
```

    ```

>        With fix:
>      ```R
>       if(is.function(value)) value <- value(levels(x))
>      ```

Reviewer: ```suggestion
       With edit (for example - change to whatever you like!):
     ```R
                           prompts = getOption("askYesNo", gettext(c("Oh yeah!", "Don't think so", "Cancel"))),
     ```
```

 5. Re-build the stats package (we only need to re-build the part we have modified). We can rebuild the package by following simple steps
    - First we need to be inside $BUILDDIR, for that we can change directory to `cd $BUILDDIR`.
    - After that we can run cmd `make` and `sudo make install` in a series to build our package again and we can now see our changes getting  reflected.

>  6. Check fix has worked as expected by re-running the reprex. Expected output without bug

Reviewer: ```suggestion
 6. Check the edit has worked as expected by re-running the example code:
```

>     ```R
>     > lm(uptake~C(Treatment, contr.treatment), CO2)
>       Call:
>       lm(formula = uptake ~ C(Treatment, contr.treatment), data = CO2)
> 
>       Coefficients:
>                          (Intercept)  C(Treatment, contr.treatment)chilled  
>                                30.64                                 -6.86 
>                                ```

Reviewer: ```suggestion
    ```R
    > askYesNo("Is this a good example?")
    Is this a good example? (Oh yeah!/don't think so/cancel) Oh yeah!
    [1] TRUE
    ```
```
